### ISSUE TITLE
`NWM` issue with fetching data for first forecast datetime, if it has `hour=0`

### ISSUE THREAD

**Issue by SorooshMani-NOAA** [schism-dev/pyschism#47]
Trying to setup hydrology in forecast mode results in errors for the first forecast if it coincides with the hour-0 forecast:
```python
from datetime import timedelta

from pyschism.forcing.source_sink import NWM
from pyschism.mesh import Gr3
from pyschism.dates import nearest_zulu

start = nearest_zulu()
rndelta = timedelta(hours=2)

gr3 = Gr3.open('MESH_PATH', crs=4326)

nwm = NWM()
nwm.write('RUNDIR', gr3, start, rndelta, overwrite=True)                                                               
```

 resulting in the following when I ran on 9/28/2022
```
Traceback (most recent call last):
  File "REPO_PATH/pyschism/forcing/source_sink/nwm.py", line 955, in write
    self._fetch_data(
  File "REPO_PATH/pyschism/forcing/source_sink/nwm.py", line 896, in _fetch_data
    nc_fid0 = Dataset(list(self.inventory.files.values())[0])["feature_id"][:]
  File "src/netCDF4/_netCDF4.pyx", line 2463, in netCDF4._netCDF4.Dataset.__init__
  File "src/netCDF4/_netCDF4.pyx", line 2026, in netCDF4._netCDF4._ensure_nc_success
FileNotFoundError: [Errno 2] No such file or directory: b'20220927/nwm.20220927/medium_range_mem1/nwm.t00z.medium_range.channel_rt_1.f024.conus.nc'
```

**Comment by SorooshMani-NOAA** [comment]
This seems to be due to the fact that `pyschism` just assigns a filename but never downloads it at:
https://github.com/schism-dev/pyschism/blob/2ecf326efa4f5a8d64b2c01179917c112f04b663/pyschism/forcing/source_sink/nwm.py#L718-L723

I'll work on a quick fix and start a pull request if it works.

**Comment by cuill** [comment]
You have to run the script to download the data for yesterday (startdate= yestarday), then re-run it again with startdate= today. It will read f024 from yesterday's folder as today's 00. NWM doesn't save hour=0 for each simulation cycle.

**Comment by SorooshMani-NOAA** [comment]
@cuill I already fixed this issue in a pull request. Please review and merge if everything looks good.

I'm aware that NWM doesn't have hour 0 forecast and we need to get 24 from the previous day, but what I meant was that if we run it with the actual start date, then it fails. Also running the script twice with two different start dates is counterintuitive. I think it makes sense for it to go through the first time we run it with the actual start date.

**Comment by cuill** [comment]
@SorooshMani-NOAA 
I have errors with your script:
botocore.exceptions.ClientError: An error occurred (404) when calling the HeadObject operation: Not Found

For each date, the self.data and self.metadata have to be updated. 
https://github.com/schism-dev/pyschism/blob/2ecf326efa4f5a8d64b2c01179917c112f04b663/pyschism/forcing/source_sink/nwm.py#L710
https://github.com/schism-dev/pyschism/blob/2ecf326efa4f5a8d64b2c01179917c112f04b663/pyschism/forcing/source_sink/nwm.py#L729

**Comment by SorooshMani-NOAA** [comment]
@cuill do you have issues using my script above or do you have issue with my updates from the pull request? In any case would it be possible for you to share the full error traceback?

**Comment by SorooshMani-NOAA** [comment]
As far as I understand my changes doesn't have any impact on `self.data` and `file_metadata` or viceversa

**Comment by cuill** [comment]
I added your updates to nwm.py and ran it. Here is the whole error information:
2022-09-27 00:00:00
[2022-09-29 11:26:12,797] pyschism.forcing.source_sink.nwm INFO: Loading pairing sources: sources.json
[2022-09-29 11:26:12,862] pyschism.forcing.source_sink.nwm INFO: Loading pairing sinks: sinks.json
[2022-09-29 11:26:13,355] pyschism.forcing.source_sink.nwm INFO: Requesting NWM data for time 2022-09-27 00:00:00
[2022-09-29 11:26:13,355] pyschism.forcing.source_sink.nwm INFO: set it=0 2022-09-27 00:00:00
[2022-09-29 11:26:13,355] pyschism.forcing.source_sink.nwm INFO: Downloading file 20220926/nwm.20220926/medium_range_mem1/nwm.t00z.medium_range.channel_rt_1.f024.conus.nc, 
Traceback (most recent call last):
  File "/sciclone/schism10/lcui01/schism20/ICOGS/ICOGS3D/Forecast/NWM/gen_sourcesink.py", line 51, in <module>
    nwm.write(output_directory, hgrid, startdate, rnday, overwrite=True)
  File "/sciclone/data10/lcui01/pyschism/pyschism/forcing/source_sink/nwm.py", line 949, in write
    self._fetch_data(
  File "/sciclone/data10/lcui01/pyschism/pyschism/forcing/source_sink/nwm.py", line 862, in _fetch_data
    self._inventory = AWSDataInventory(
  File "/sciclone/data10/lcui01/pyschism/pyschism/forcing/source_sink/nwm.py", line 727, in __init__
    self.files[requested_time] = self._download_forecast_ifneeded(key)
  File "/sciclone/data10/lcui01/pyschism/pyschism/forcing/source_sink/nwm.py", line 739, in _download_forecast_ifneeded
    self.s3.download_fileobj(self.bucket, s3_key, f)
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/boto3/s3/inject.py", line 795, in download_fileobj
    return future.result()
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/s3transfer/futures.py", line 103, in result
    return self._coordinator.result()
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/s3transfer/futures.py", line 266, in result
    raise self._exception
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/s3transfer/tasks.py", line 269, in _main
    self._submit(transfer_future=transfer_future, **kwargs)
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/s3transfer/download.py", line 354, in _submit
    response = client.head_object(
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/botocore/client.py", line 508, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/sciclone/home20/lcui01/anaconda3/envs/hotstart2/lib/python3.9/site-packages/botocore/client.py", line 915, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (404) when calling the HeadObject operation: Not Found

Yes, your changes didn't impact self.data and self.file_metadata, that is the reason why it failed to download the file for hour 0, which is f024 from the day before's folder. You have to update self.data to download f024. Because AWSForecastInvetory  only works for forecast, which is assumed to continuously run and getting f024 from pre-downloaded files is not an issue. This works for pyschism version of STOFS3D forecast on Frontera.

**Comment by SorooshMani-NOAA** [comment]
@cuill it's a strange issue. Let me see if I can do something to get this error somehow, because right now it works fine for me. Are you running my script by only replacing mesh and output paths?

What's strange to me is that the run date of the script is today (9/29) but yesterday date is 9/27 in the error message. How's that possible?

About your comments:
> Yes, your changes didn't impact self.data and self.file_metadata, that is the reason why it failed to download the file for hour 0, which is f024 from the day before's folder.

I still don't see why `self.data` or `file_metadata` needs to be updated; the `_download_forecast_ifneeded` method doesn't use either of these variables.

https://github.com/schism-dev/pyschism/pull/48/files#diff-a37f0e3734a833b908a7c22ef0a4a419e94b9d3fa4efc475ce46e7e7f3c83aabR727

Also the error you get seem to be related to a file not being available for download! Can you print the key

> Because AWSForecastInvetory only works for forecast, which is assumed to continuously run and getting f024 from pre-downloaded files is not an issue. This works for pyschism version of STOFS3D forecast on Frontera

In my opinion the fact that one needs to run the `NWM.write` code multiple times, once with a fake start date and once with an actual simulation start date to get the forecast data means there's *usability* issue. The fact that it works fine for STOFS3D doesn't show there are no issues. In fact it shows the code didn't consider generic use-case; for example, the forecast should not be assumed to be running continuously, unless `pyschism` was developed only for STOFS3D for example. 

This also breaks using the `NWM` forcing with `ModelConfig`, which was the original intention behind `pyschism`

**Comment by cuill** [comment]
@SorooshMani-NOAA 
The issue is fixed. Please pull the latest version.

**Comment by SorooshMani-NOAA** [comment]
Thank you, I'll test with the updates

**Comment by SorooshMani-NOAA** [comment]
@cuill thanks it works successfully generates the nwm inputs now. The only issue I see is that instead of using temporary dir it now keeps the downloaded files on the current directory.

**Comment by cuill** [comment]
@SorooshMani-NOAA Yes, we would like to keep the files in case it is needed to reproduce the run. It can be modified to be an option.

**Comment by SorooshMani-NOAA** [comment]
I close this ticket since the main issue was resolved by your change at [96e52fd](https://github.com/schism-dev/pyschism/commit/96e52fd54ff9beacc5e30e7fe2821989d0b67bd4)
